File: scala/api/src/main/scala/com/tersesystems/echopraxia/sapi/support/AsyncLoggerMethods.scala
package com.tersesystems.echopraxia.sapi.support

import com.tersesystems.echopraxia.{Field, LoggerHandle}
import com.tersesystems.echopraxia.sapi.Condition

trait AsyncLoggerMethods[FB] {

  // ------------------------------------------------------------------------
  // TRACE

  /**
   * Logs using a logger handle at TRACE level.
   *
   * @param consumer the consumer of the logger handle.
   */
  def trace(consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs using a condition and a logger handle at TRACE level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  def trace(c: Condition, consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs statement at TRACE level.
   *
   * @param message the given message.
   */
  def trace(message: String): Unit

  /**
   * Logs statement at TRACE level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  def trace(message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Logs statement at TRACE level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  def trace(message: String, e: Throwable): Unit

  /**
   * Conditionally logs statement at TRACE level.
   *
   * @param condition the given condition.
   * @param message   the message.
   */
  def trace(condition: Condition, message: String): Unit

  /**
   * Conditionally logs statement at TRACE level using a field builder function.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param f         the field builder function.
   */
  def trace(condition: Condition, message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Conditionally logs statement at TRACE level with exception.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param e         the given exception.
   */
  def trace(condition: Condition, message: String, e: Throwable): Unit

  // ------------------------------------------------------------------------
  // DEBUG

  /**
   * Logs using a logger handle at DEBUG level.
   *
   * @param consumer the consumer of the logger handle.
   */
  def debug(consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs using a condition and a logger handle at DEBUG level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  def debug(c: Condition, consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs statement at DEBUG level.
   *
   * @param message the given message.
   */
  def debug(message: String): Unit

  /**
   * Logs statement at DEBUG level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  def debug(message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Logs statement at DEBUG level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  def debug(message: String, e: Throwable): Unit

  /**
   * Conditionally logs statement at DEBUG level.
   *
   * @param condition the given condition.
   * @param message   the message.
   */
  def debug(condition: Condition, message: String): Unit

  /**
   * Conditionally logs statement at DEBUG level using a field builder function.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param f         the field builder function.
   */
  def debug(condition: Condition, message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Conditionally logs statement at DEBUG level with exception.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param e         the given exception.
   */
  def debug(condition: Condition, message: String, e: Throwable): Unit

  // ------------------------------------------------------------------------
  // INFO

  /**
   * Logs using a logger handle at INFO level.
   *
   * @param consumer the consumer of the logger handle.
   */
  def info(consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs using a condition and a logger handle at INFO level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  def info(c: Condition, consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs statement at INFO level.
   *
   * @param message the given message.
   */
  def info(message: String): Unit

  /**
   * Logs statement at INFO level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  def info(message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Logs statement at INFO level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  def info(message: String, e: Throwable): Unit

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message   the message.
   */
  def info(condition: Condition, message: String): Unit

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param f         the field builder function.
   */
  def info(condition: Condition, message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message   the message.
   * @param e         the given exception.
   */
  def info(condition: Condition, message: String, e: Throwable): Unit

  // ------------------------------------------------------------------------
  // WARN

  /**
   * Logs using a logger handle at WARN level.
   *
   * @param consumer the consumer of the logger handle.
   */
  // WARN
  def warn(consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs using a condition and a logger handle at WARN level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  def warn(c: Condition, consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs statement at WARN level.
   *
   * @param message the given message.
   */
  def warn(message: String): Unit

  /**
   * Logs statement at WARN level using a field builder function.
   *
   * @param message the message.
   * @param f       the field builder function.
   */
  def warn(message: String, f: FB => java.util.List[Field]): Unit

  /**
   * Logs statement at WARN level with exception.
   *
   * @param message the message.
   * @param e       the given exception.
   */
  def warn(message: String, e: Throwable): Unit

  def warn(condition: Condition, message: String): Unit

  def warn(condition: Condition, message: String, f: FB => java.util.List[Field]): Unit

  def warn(condition: Condition, message: String, e: Throwable): Unit

  // ------------------------------------------------------------------------
  // ERROR

  /**
   * Logs using a logger handle at ERROR level.
   *
   * @param consumer the consumer of the logger handle.
   */
  def error(consumer: LoggerHandle[FB] => Unit): Unit

  /**
   * Logs using a condition and a logger handle at ERROR level.
   *
   * @param c        the condition
   * @param consumer the consumer of the logger handle.
   */
  def error(c: Condition, consumer: LoggerHandle[FB] => Unit): Unit

  def error(message: String): Unit

  def error(message: String, f: FB => java.util.List[Field]): Unit

  def error(message: String, e: Throwable): Unit

  def error(condition: Condition, message: String): Unit

  def error(condition: Condition, message: String, f: FB => java.util.List[Field]): Unit

  def error(condition: Condition, message: String, e: Throwable): Unit
}

File: async/src/main/java/com/tersesystems/echopraxia/async/support/AsyncLoggerMethods.java
package com.tersesystems.echopraxia.async.support;

import com.tersesystems.echopraxia.Condition;
import com.tersesystems.echopraxia.Field;
import com.tersesystems.echopraxia.LoggerHandle;
import java.util.function.Consumer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Logging method API used by the AsyncLogger.
 *
 * @param <FB> the field builder type.
 */
public interface AsyncLoggerMethods<FB> {

  void trace(@NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs using a condition and a logger handle at TRACE level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  void trace(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs statement at TRACE level.
   *
   * @param message the given message.
   */
  void trace(@Nullable String message);

  /**
   * Logs statement at TRACE level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  void trace(@Nullable String message, Field.@NotNull BuilderFunction<FB> f);

  /**
   * Logs statement at TRACE level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  void trace(@Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at TRACE level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  void trace(@NotNull Condition condition, @Nullable String message);

  /**
   * Conditionally logs statement at TRACE level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  void trace(
      @NotNull Condition condition, @Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Conditionally logs statement at TRACE level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  void trace(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e);

  /**
   * Logs statement at DEBUG level.
   *
   * @param message the given message.
   */
  void debug(@Nullable String message);

  /**
   * Logs statement at DEBUG level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  void debug(@Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Logs statement at DEBUG level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  void debug(@Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at DEBUG level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  void debug(@NotNull Condition condition, @Nullable String message);

  /**
   * Conditionally logs statement at DEBUG level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  void debug(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at DEBUG level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  void debug(
      @NotNull Condition condition, @Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Logs using a logger handle at DEBUG level.
   *
   * @param consumer the consumer of the logger handle.
   */
  void debug(@NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs using a condition and a logger handle at DEBUG level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  void debug(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs statement at INFO level.
   *
   * @param message the given message.
   */
  void info(@Nullable String message);

  /**
   * Logs statement at INFO level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  void info(@Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Logs statement at INFO level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  void info(@Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  void info(@NotNull Condition condition, @Nullable String message);

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  void info(
      @NotNull Condition condition, @Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  void info(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e);

  /**
   * Logs using a condition and a logger handle at INFO level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  void info(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs statement at WARN level.
   *
   * @param message the given message.
   */
  void warn(@Nullable String message);

  /**
   * Logs statement at WARN level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  void warn(@Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Logs statement at WARN level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  void warn(@Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  void warn(@NotNull Condition condition, @Nullable String message);

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  void warn(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  void warn(
      @NotNull Condition condition, @Nullable String message, @NotNull Field.BuilderFunction<FB> f);
  /**
   * Logs using a logger handle at WARN level.
   *
   * @param consumer the consumer of the logger handle.
   */
  void warn(@NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs using a condition and a logger handle at WARN level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  void warn(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs statement at INFO level.
   *
   * @param message the given message.
   */
  void error(@Nullable String message);

  /**
   * Logs statement at INFO level using a field builder function.
   *
   * @param message the message.
   * @param f the field builder function.
   */
  void error(@Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Logs statement at INFO level with exception.
   *
   * @param message the message.
   * @param e the given exception.
   */
  void error(@Nullable String message, @NotNull Throwable e);

  /**
   * Conditionally logs statement at INFO level.
   *
   * @param condition the given condition.
   * @param message the message.
   */
  void error(@NotNull Condition condition, @Nullable String message);

  /**
   * Conditionally logs statement at INFO level using a field builder function.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param f the field builder function.
   */
  void error(
      @NotNull Condition condition, @Nullable String message, @NotNull Field.BuilderFunction<FB> f);

  /**
   * Conditionally logs statement at INFO level with exception.
   *
   * @param condition the given condition.
   * @param message the message.
   * @param e the given exception.
   */
  void error(@NotNull Condition condition, @Nullable String message, @NotNull Throwable e);

  /**
   * Logs using a logger handle at ERROR level.
   *
   * @param consumer the consumer of the logger handle.
   */
  void error(@NotNull Consumer<LoggerHandle<FB>> consumer);

  /**
   * Logs using a condition and a logger handle at ERROR level.
   *
   * @param c the condition
   * @param consumer the consumer of the logger handle.
   */
  void error(@NotNull Condition c, @NotNull Consumer<LoggerHandle<FB>> consumer);
}
